src/app.tsx

```diff
@@ -0,0 +1,105 @@
+// tslint:disable-next-line:no-implicit-dependencies
+import { Opportunity } from "@src/generated/sobs";
+import { Affix, Button, Col, Icon, Layout, Row, Spin } from "antd";
+import * as React from "react";
+
+interface Props {
+    id: string;
+}
+
+interface State {
+    isOpptyLoading: boolean;
+    isTouchSaving: boolean;
+    oppName: string;
```

Comment:
you could just store the `OpportunityFeilds` directly on the state and then when you need to update state, use [ES6 spread syntax.](https://callawaycloud.my.salesforce.com/a0K1C00000JYtS5) to do so without mutating:

```
let newOppty = {...this.state.oppty, ...{touches: newCount}};
await new Opportunity(newOppty).update(); 
this.setState({oppty: newOppty});
```

Reply:
so much cool syntax

in your experience has storing the full sobject (the immutable fields instance) worked out better than storing just the data points you need? seems like it would be less to manage, but could invite extra functions you didn't intend (or maybe that's a positive since you don't need to adjust your state logic every time you want to start referring to a new field on a page)

Reply:
@ralphcallaway  that's a great question.

There's definitely a trade off to each.   It's MUCH easier and efficient (productivity wise) to just store the entire object.  When you are rapidly prototyping, it makes things come together really fast.

However,  it violates Idiomatic redux for two reasons:

1: the SObjects returned by ts-force are classes and have functions, which apparently can have negative impacts on performance (you are only suppose to store POJOS).

2: ts-force allows you to pull dependency trees down and they are all stored on the object.  This violates the best practice of normalizing your reducers.  You could easily use a library like [normalizr](https://github.com/paularmstrong/normalizr) to solve this issue

My approach has been to follow this simple rule:

> It's ok to store SObjects in the reducer ONLY when you don't need to mutate the objects themselves.  It's probably not a huge deal to change first class properties, but definitely shouldn't change children objects

The only downside has been that my state size has gotten really big, due to duplicate relationship data being pulled down.  For example, I'm storing treatments on every Spotlight Product.  There might be 5000 products in my store, each with object data, but there may only be 15 distinct treatments across all of them.

I don't think this has a big impact on rendering performance, but it does make things like trying to store state in Saleforce field harder (I currently need 10 long text fields to be safe!).